### PR TITLE
Add unit  test  for RebaseArchiveEntries

### PR DESCRIPTION
**Pull request by keloyang**
Add testcase for https://github.com/moby/moby/pull/33956.

### CODE REVIEW

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]
Please add API test instead.

Also, latest CLI no longer works with integration-cli: https://github.com/docker/cli/issues/293

**Comment by keloyang** [comment]
@AkihiroSuda, thanks for  your advice, but I still don't know how to change, could you give me a example ?

**Comment by keloyang** [comment]
@AkihiroSuda thanks,  where to add API test ?

**Comment by AkihiroSuda** [comment]
docker_api_*_test.go

**Comment by keloyang** [comment]
@AkihiroSuda  I  have update the code to used API test instead, PTAL.

File: integration-cli/docker_cli_cp_utils_test.go [hunk 1]
```diff
@@ -103,8 +108,8 @@ func makeTestContentInDir(c *check.C, dir string) {
 		case ftSymlink:
 			c.Assert(os.Symlink(fd.contents, path), checker.IsNil)
 		}
-
-		if fd.filetype != ftSymlink && runtime.GOOS != "windows" {
+		// Need skip Hardlink
```
Comment (dnephin) [inline integration-cli/docker_cli_cp_utils_test.go]: I don't think this comment provides any value over reading the code. It's actually confusing because the code skips both kinds of links, but the comment only talks about hard links.

File: integration-cli/docker_api_containers_test.go [hunk 2]
```diff
@@ -1948,3 +1949,27 @@ func (s *DockerSuite) TestContainerKillCustomStopSignal(c *check.C) {
 	err = waitInspect(id, "{{.State.Running}} {{.State.Restarting}}", "false false", 30*time.Second)
 	c.Assert(err, checker.IsNil)
 }
+
+func (s *DockerSuite) TestCpAPIWithHardlink(c *check.C) {
```
Comment (dnephin) [inline integration-cli/docker_api_containers_test.go]: The API is called `archive` not `Co`, so I think this test should be called:  `TestContainerAPIArchiveWithHardlink`.

It looks like a bunch of the tests in this files aren't following that pattern, but I believe that's the original pattern.

File: integration-cli/docker_api_containers_test.go [hunk 3]
```diff
@@ -1948,3 +1949,27 @@ func (s *DockerSuite) TestContainerKillCustomStopSignal(c *check.C) {
...
+	res, body, err := request.SockRequestRaw("GET", fmt.Sprintf("/containers/%s/archive?path=%s", containerID, srcPath), nil, "application/x-tar", daemonHost())
+	c.Assert(err, checker.IsNil)
+	c.Assert(res.StatusCode, checker.Equals, http.StatusOK)
+
+	srcInfo := archive.CopyInfo{
+		Path:   srcPath,
+		Exists: true,
+		IsDir:  true,
+	}
+	archive.CopyTo(body, srcInfo, dstPath)
```
Comment (dnephin) [inline integration-cli/docker_api_containers_test.go]: I don't think I understand why this is copying here. Shouldn't the test just assert that the file is where it is expected to be?  I think you could stat both and compare the `os.FileInfo` ?

**Comment by aaronlehmann** [comment]
Can you please update the title of the PR?

File: integration-cli/docker_api_containers_test.go (hunk 3 above)
Comment (keloyang) [inline integration-cli/docker_api_containers_test.go]: It need archive.Copy to make the file is where it is expected to be,  and if the two files have the same content, the file is sure to exist.

File: integration-cli/docker_api_containers_test.go [hunk 4]
```diff
@@ -1948,3 +1949,28 @@ func (s *DockerSuite) TestContainerKillCustomStopSignal(c *check.C) {
...
+	res, body, err := request.SockRequestRaw("GET", fmt.Sprintf("/containers/%s/archive?path=%s", containerID, srcPath), nil, "application/x-tar", daemonHost())
+	c.Assert(err, checker.IsNil)
+	c.Assert(res.StatusCode, checker.Equals, http.StatusOK)
+
+	srcInfo := archive.CopyInfo{
+		Path:   srcPath,
+		Exists: true,
+		IsDir:  true,
+	}
+	err = archive.CopyTo(body, srcInfo, dstPath)
```
Comment (dnephin) [inline integration-cli/docker_api_containers_test.go]: Isn't the copy already performed by the archive call?

Comment (keloyang) [inline integration-cli/docker_api_containers_test.go]: After "SockRequestRaw", it only get a io.ReadCloser, and we need use archive.CopyTo to read tar data and untar it to dstPath.

Comment (dnephin) [inline integration-cli/docker_api_containers_test.go]: I see. Would `archive.Untar()` work here as well? That seems like it would make the intent more obvious.

File: integration-cli/docker_api_containers_test.go [hunk 5]
```diff
@@ -1948,3 +1949,28 @@ func (s *DockerSuite) TestContainerKillCustomStopSignal(c *check.C) {
...
+	c.Assert(err, checker.IsNil)
+	c.Assert(fileContentEquals(c, cpPath(tmpDir, "dir6_/hardLinkFile1"), "dir6/hardLinkFile1\n"), checker.IsNil)
```
Comment (dnephin) [inline integration-cli/docker_api_containers_test.go]: Is there a more strict assertion we can use? Could we compare the `os.FileInfo` for these files?

Comment (keloyang) [inline integration-cli/docker_api_containers_test.go]: this code test the following case,
```
docker cp id:/dir6  /tmpxxx/_dir6
```
"/dir6_/hardLinkFile1" is in container, we can't use it to compare. My goal is testing the issue located in https://github.com/moby/moby/pull/33956,and if the file have the exact content, I think we don't need compare  os.FileInfo ,  -_-  .

File: integration-cli/docker_api_containers_test.go (hunk 4 above)
Comment (keloyang) [inline integration-cli/docker_api_containers_test.go]: If use  archive.Untar, I still need add some code like CopyInfoDestinationPath and PrepareArchiveCopy and then use archive.Untar,  so use CopyTo is much clearer.

**Comment by dnephin** [comment]
> this code test the following case,
> docker cp id:/dir6  /tmpxxx/_dir6

It can't test that case because the client is not part of this repo. It's only testing the API endpoint.

If the API endpoint returns a tar archive, all we need to assert is that the contents of the tar archive are as we expect. That's why I don't really understand why we need `CopyTo`.

If FileInfo isn't the right way to assert. We could instead just assert the tar archive has the expected file  with the correct header. 

I don't think checking content is correct because the file could be a copy (not a hard link), or a symlink and still have the same content, right?

**Comment by keloyang** [comment]
@dnephin  thanks for your patient reply, updated, PTAL.

File: integration-cli/docker_api_containers_unix_test.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+	dstPath := cpPath(tmpDir, "dir6_")
+	dstFile := cpPath(tmpDir, "dir6_/hardLinkFile1")
+	endpoint := fmt.Sprintf("/containers/%s/archive?path=%s", containerID, srcPath)
+	res, body, err := request.SockRequestRaw("GET", endpoint, nil, "application/x-tar", daemonHost())
+	c.Assert(err, checker.IsNil)
+	c.Assert(res.StatusCode, checker.Equals, http.StatusOK)
+
+	dstDir, dstBase := archive.SplitPathDirEntry(dstPath)
+	_, srcBase := archive.SplitPathDirEntry(srcPath)
+	copyArchive := archive.RebaseArchiveEntries(body, srcBase, dstBase)
```
Comment (dnephin) [inline integration-cli/docker_api_containers_unix_test.go]: Sorry, why is this necessary? Don't we already know the expected path for the file?

Comment (keloyang) [inline integration-cli/docker_api_containers_unix_test.go]: https://github.com/moby/moby/pull/33956 add the following three line, and the code is in RebaseArchiveEntries.
```
+++ b/pkg/archive/copy.go
@@ -332,6 +332,9 @@ func RebaseArchiveEntries(srcContent io.Reader, oldBase, newBase string) io.Read
                        }
 
                        hdr.Name = strings.Replace(hdr.Name, oldBase, newBase, 1)
+                       if hdr.Typeflag == tar.TypeLink {
+                               hdr.Linkname = strings.Replace(hdr.Linkname, oldBase, newBase, 1)
+                       }
 
                        if err = rebasedTar.WriteHeader(hdr); err != nil {
                                w.CloseWithError(err)
```

Comment (dnephin) [inline integration-cli/docker_api_containers_unix_test.go]: Sorry, I wasn't clear. I meant to ask, why is `RebaseArchiveEntries()` being called here?

It seems to just rename files inside the archive, but I don't understand why that's necessary for this test. We should be just asserting that the archive contains the files we expect.

I'd like to suggest another approach. Instead of making this test about just hardlinks, we should make this a general purpose test for the archive endpoint.

It looks like we have a good way of creating a container with a bunch of different types of files (`makeTestContainer()`), so all we need is to assert that all the files in the archive are as we expect and we can test everything with this one case.

Does that make sense?

Comment (keloyang) [inline integration-cli/docker_api_containers_unix_test.go]: @dnephin thanks for your reply, I 'a not sure whether I understand you completely, Let me show what I think.

> docker cp id:/dir6 /tmpxxx/_dir6

the case has the following feathres,
- there is hardlink in  id:/dir6 
- the dir "_dir6" is not exist.
- the baseDirs(dir6 and _dir6) are not the same, old is dir6, new is dir_, dir6 != _dir6

before https://github.com/moby/moby/pull/33956, RebaseArchiveEntries has a bug, it don't change base name, it will lead the following issue,

"link /tmp/test-cp-from-dir-with-hardlink085642536/**dir6**/hardFile1 /tmp/test-cp-from-dir-with-hardlink085642536/**dir6_**/hardLinkFile1: no such file or directory"

the case above can reproduce the issue before https://github.com/moby/moby/pull/33956.

the correct should be,
"link /tmp/test-cp-from-dir-with-hardlink085642536/**dir6_**/hardFile1 /tmp/test-cp-from-dir-with-hardlink085642536/**dir6_**/hardLinkFile1"

and https://github.com/moby/moby/pull/33956 fix it. So RebaseArchiveEntries()  is called here.

Comment (dnephin) [inline integration-cli/docker_api_containers_unix_test.go]: I see, thank you for clarifying this!

I don't think this should be an integration test at all. We can test `RebaseArchiveEntries` without the daemon. Can you please make this a unit test?

The unit test can create an archive with a hardlink, and then call `RebaseArchiveEntries()` and check that the archive is correct afterward.

**Comment by keloyang** [comment]
@dnephin updated.

File: pkg/archive/archive_unix_test.go [hunk 7]
```diff
@@ -126,6 +126,59 @@ func TestTarWithHardLink(t *testing.T) {
 	}
 }
 
+func TestTarWithHardLinkAndRebase(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "docker-test-tar-hardlink-rebase")
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (dnephin) [inline pkg/archive/archive_unix_test.go]:
```go
require.NoError(t, err)
```

File: pkg/archive/archive_unix_test.go [hunk 8]
```diff
@@ -126,6 +126,59 @@ func TestTarWithHardLink(t *testing.T) {
...
+
+	if i1, err = getInode(filepath.Join(dest, "1")); err != nil {
+		t.Fatal(err)
+	}
+	if i2, err = getInode(filepath.Join(dest, "2")); err != nil {
+		t.Fatal(err)
+	}
+
+	if i1 != i2 {
```
Comment (dnephin) [inline pkg/archive/archive_unix_test.go]:
```go
assert.Equal(t, i1, i2)
```

**Review by dnephin — APPROVED** [review]
Thanks! this looks much better

LGTM

but using the assertions package would be good too

File: pkg/archive/archive_unix_test.go (hunk 8 above)
Comment (keloyang) [inline pkg/archive/archive_unix_test.go]: I think so.

**Review by AkihiroSuda — APPROVED** [review]
LGTM if green

**Review by yongtang — APPROVED** [review]
LGTM and all Jenkins tests passed.
